src/compiler/checker.ts

```diff
@@ -16173,7 +16173,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function getSubstitutionType(baseType: Type, constraint: Type) {
-        return constraint.flags & TypeFlags.AnyOrUnknown || constraint === baseType || baseType.flags & TypeFlags.Any ?
+        return constraint.flags & TypeFlags.AnyOrUnknown || constraint === baseType || baseType.flags & TypeFlags.Any || !isGenericType(constraint) && !isGenericType(baseType) && isTypeAssignableTo(baseType, constraint) ?

```

Comment:
alternatively, substitution types could get some improved handling in `getPropertyTypeForIndexType`

Reply:
Yeah, I think I'd prefer `getPropertyNameFromIndex` + `getPropertyTypeForIndexType` being updated to handle substitutions - the more places we recognize them as transparently their original type (if the substituted version isn't useful), the better, since it makes it easier on us to broaden the scope of how we use substitution types to support other features.

As far as this approach goes... `instantiateTypeWorker` has similar logic to this - honestly, a lot of that logic should *probably* get moved into here, so it happens on initial construction *and* instantiation, rather than just instantiation.

Reply:
I agree, the right approach is to erase them where needed. Omitting the substitution type just because the constraint is assignable isn't right and will definitely have adverse effects as the constraint may be a more specific type.

Reply:
I pushed out the requested change. I decided to use a similar approach of unwrapping the substation type early in `getPropertyTypeForIndexType` like it was done in https://github.com/microsoft/TypeScript/pull/58292